src/gebt_poc/solver.cpp

```diff
@@ -16,22 +16,46 @@ UserDefinedQuadrature::UserDefinedQuadrature(
 }
 
 void Interpolate(
+    Kokkos::View<double*> nodal_values, Kokkos::View<double*> interpolation_function,
+    Kokkos::View<double*> interpolated_values
+) {
+    const auto n_nodes = nodal_values.extent(0) / kNumberOfLieAlgebraComponents;
+    KokkosBlas::fill(interpolated_values, 0.);
+    for (std::size_t i = 0; i < n_nodes; ++i) {
+        auto index = i * kNumberOfLieAlgebraComponents;
+        KokkosBlas::axpy(
+            interpolation_function(i),
+            Kokkos::subview(
+                nodal_values, Kokkos::pair(index, index + kNumberOfLieAlgebraComponents)
+            ),
+            interpolated_values
+        );
+    }
+    // Normalize the rotation quaternion
+    auto q = Kokkos::subview(interpolated_values, Kokkos::pair(3, 7));
+    auto norm = KokkosBlas::nrm2(q);
+    if (norm != 0.0) {
+        KokkosBlas::scal(q, 1. / norm, q);
+    }
+}
+
+void InterpolateDerivative(
     Kokkos::View<double*> nodal_values, Kokkos::View<double*> interpolation_function,
     const double jacobian, Kokkos::View<double*> interpolated_values
 ) {
     const auto n_nodes = nodal_values.extent(0) / kNumberOfLieAlgebraComponents;
-    Kokkos::deep_copy(interpolated_values, 0.);
-    for (std::size_t i = 0; i < kNumberOfLieAlgebraComponents; ++i) {
-        // TODO How to remove this parallel_reduce?
-        Kokkos::parallel_reduce(
-            n_nodes,
-            KOKKOS_LAMBDA(const size_t j, double& value) {
-                value += interpolation_function(j) *
-                         nodal_values(j * kNumberOfLieAlgebraComponents + i) / jacobian;
-            },
-            Kokkos::Sum(Kokkos::subview(interpolated_values, i))
+    KokkosBlas::fill(interpolated_values, 0.);
+    for (std::size_t i = 0; i < n_nodes; ++i) {
+        auto index = i * kNumberOfLieAlgebraComponents;
+        KokkosBlas::axpy(
+            interpolation_function(i),
+            Kokkos::subview(
+                nodal_values, Kokkos::pair(index, index + kNumberOfLieAlgebraComponents)
+            ),
+            interpolated_values
```

Comment:
Similar comments to the equivalent lines in the Interpolate function

Reply:
I've changed the function signature so that the data remains on the host.